File: F28/VB/2.018/VBF28LightControlDemo/EasyAutoLearning.vb

Imports VBF28LightControlDemo.F28LightControl


Public Class EasyAutoLearning


    ' Auto-Ratio phase
    Enum EASY_AUTO_PHASES As Byte
        EASY_AUTO_IDDLE
        EASY_AUTO_START
        EASY_AUTO_WAIT_EOC
        EASY_AUTO_END
    End Enum


    Dim m_wNbCycles, m_wInterCycle As UShort

    Dim m_wError As UShort
    Dim m_ucPhase As Byte
    Dim m_sModuleId As Short
    Dim m_bRunning As Boolean

    ' *****************************************************

    '
    ' *****************************************************
    ' Iniialize
    ' *****************************************************
    '
    Public Sub Initialize()
        m_wNbCycles = 2
        m_wInterCycle = 3000          ' 3 sec

        m_wError = 0
        m_sModuleId = 0
        m_bRunning = False

        m_ucPhase = EASY_AUTO_PHASES.EASY_AUTO_IDDLE

    End Sub
    '
    ' *****************************************************
    ' Is Running
    ' *****************************************************
    '
    Public Function IsRunning() As Boolean

        Return m_bRunning

    End Function

    '
    ' *****************************************************
    ' Get Error
    ' *****************************************************
    '
    Public Function GetError() As UShort

        Return m_wError

    End Function

    '
    ' *****************************************************
    ' Get phase
    ' *****************************************************
    '
    Public Function GetPhase() As Byte

        Return m_ucPhase

    End Function

    '
    ' *****************************************************
    ' Get Calibration phase 
    ' *****************************************************
    '
    Public Function GetPhaseStr(ucPhase As Byte) As String

        Dim strRet As String

        strRet = "????"
        Select Case ucPhase

            Case EASY_AUTO_PHASES.EASY_AUTO_IDDLE
                strRet = "Ready"

            Case EASY_AUTO_PHASES.EASY_AUTO_START
                strRet = "Start Auto-Ratio"

            Case EASY_AUTO_PHASES.EASY_AUTO_WAIT_EOC
                strRet = "Wait end of Easy Auto Learning calculation"

            Case EASY_AUTO_PHASES.EASY_AUTO_END
                strRet = "End"

        End Select

        Return strRet

    End Function
    '
    ' *****************************************************
    ' Read alarm code at the end of calibration
    ' *****************************************************
    '
    Public Function GetAlarm() As UShort
        Dim wAlarm As UShort

        wAlarm = 0

        If m_sModuleId > 0 Then

            wAlarm = F28_GetAutoCalAlarm(m_sModuleId)

        End If

        Return wAlarm

    End Function

    '
    ' *****************************************************
    ' Start Calibration
    ' *****************************************************
    '
    Public Function StartEasyAutoLearning(sModuleID As Short, _
                           wNbCycles As UShort, _
                           wInterCycle As UShort) As Boolean
        Dim bRet As Boolean

        bRet = False

        If sModuleID > 0 Then

            m_wNbCycles = wNbCycles
            m_wInterCycle = wInterCycle

            m_sModuleId = sModuleID
            m_wError = 0
            m_ucPhase = EASY_AUTO_PHASES.EASY_AUTO_START
            m_bRunning = True

            bRet = True

        End If

        Return bRet

    End Function

    '
    ' *****************************************************
    ' Stop Auto-ration
    ' *****************************************************
    '
    Public Function StopEasyAutoLearning() As Short

        Dim sRet As Short

        sRet = F28_RETURN.F28_FAIL

        If (m_sModuleId > 0) And (m_ucPhase <> EASY_AUTO_PHASES.EASY_AUTO_IDDLE) Then

            sRet = F28_StopEasyAutoLearning(m_sModuleId)

        End If

        m_ucPhase = EASY_AUTO_PHASES.EASY_AUTO_IDDLE
        m_bRunning = False

        Return sRet

    End Function

    '
    ' *****************************************************
    ' Purpose   : Run Auto-Ration
    ' Return    :
    '               - True  : EOC calibration
    '               - False : Running
    ' *****************************************************
    '
    Public Function RunEasyAutoLearning() As Boolean
        Dim sRet As Short
        Dim bReturn As Boolean

        ' Not End of Run
        bReturn = False

        Select Case m_ucPhase

            Case EASY_AUTO_PHASES.EASY_AUTO_START
                sRet = F28_StartEasyAutoLearning(m_sModuleId, m_wNbCycles, m_wInterCycle)

                If (sRet = F28_RETURN.F28_OK) Then
                    m_ucPhase = EASY_AUTO_PHASES.EASY_AUTO_WAIT_EOC
                Else
                    m_wError = m_ucPhase
                    m_ucPhase = EASY_AUTO_PHASES.EASY_AUTO_END
                End If

            Case EASY_AUTO_PHASES.EASY_AUTO_WAIT_EOC           ' Wait EOC 

                If (F28_GetEOCEasyAutoLearning(m_sModuleId) > 0) Then
                    m_wError = 0        ' Pas d'erreur
                    m_ucPhase = EASY_AUTO_PHASES.EASY_AUTO_END
                End If

            Case EASY_AUTO_PHASES.EASY_AUTO_END                       ' End of auto Auto-Ratio
                m_wError = m_ucPhase
                m_ucPhase = EASY_AUTO_PHASES.EASY_AUTO_IDDLE
                m_bRunning = False
                bReturn = True

            Case EASY_AUTO_PHASES.EASY_AUTO_IDDLE                     ' Ready do nothing

                ' do nothing

        End Select

        Return bReturn

    End Function

End Class

File: F28/VC/2.018/VCF28LightControlDemo/EasyAutoLearning.cpp

#include "stdafx.h"
#include <F28LightControl_ETH.h>
#include "EasyAutoLearning.h"

CEasyAutoLearning::CEasyAutoLearning()
{
	Initialize();
}

CEasyAutoLearning::~CEasyAutoLearning()
{
}

void CEasyAutoLearning::Initialize(void)
{
	m_wNbCycles		= 2;
	m_wInterCycle	= 3000;		// 3 sec
	m_wError		= 0;
	m_sModuleId		= 0;
	m_bRunning		= FALSE;
	m_ucPhase		= EASY_AUTO_IDDLE; 
}

CString CEasyAutoLearning::GetPhaseStr(BYTE ucPhase)
{
	CString strRet = _T("????");
	strRet.Format(_T("<%d>: "), ucPhase);

	switch (ucPhase)
	{
	case EASY_AUTO_IDDLE:
		strRet += _T("Ready");
		break;

	case EASY_AUTO_START:
		strRet += _T("Start Easy Auto Learning calculation");
		break;

	case EASY_AUTO_WAIT_EOC:
		strRet += _T("Wait end of Easy Auto Learning calculation");
		break;

	case EASY_AUTO_END:
		strRet += _T("End Easy Auto Learning");
		break;
	}

	return strRet;
}

BYTE CEasyAutoLearning::GetAlarm()
{
	BYTE ucAlarm = 0;
	if (m_sModuleId > 0)
	{
		ucAlarm = F28_GetAutoCalAlarm(m_sModuleId);
	}
	return ucAlarm;
}

BOOL CEasyAutoLearning::StartEasyAutoLearning(short sModuleID, WORD wNbCycles, WORD wInterCycle)
{
	BOOL bRet = FALSE;

	if (sModuleID > 0)
	{
		m_wNbCycles = wNbCycles;
		m_wInterCycle = wInterCycle;

		m_sModuleId = sModuleID;
		m_wError = 0;
		m_ucPhase = EASY_AUTO_START;
		m_bRunning = TRUE;

		bRet = TRUE;
	}

	return bRet;
}

BOOL CEasyAutoLearning::StopEasyAutoLearning()
{
	short sRet = F28_FAIL;

	if ((m_sModuleId > 0) && (m_ucPhase != EASY_AUTO_IDDLE))
	{
		sRet = F28_StopEasyAutoLearning(m_sModuleId);
	}

	m_ucPhase = EASY_AUTO_IDDLE;
	m_bRunning = FALSE;

	return (sRet == F28_OK) ? TRUE:FALSE;
}

//' *****************************************************
//' Purpose   : Run Easy Auto Learning
//' Return    :
//'               - True  : EOC Easy auto learning
//'               - False : Running
//' *****************************************************
BOOL CEasyAutoLearning::RunEasyAutoLearning()
{
	short sRet;

	BOOL bReturn = FALSE;

	switch (m_ucPhase)
	{
	case EASY_AUTO_START:			// Start Easy auto learning
		sRet = F28_StartEasyAutoLearning(m_sModuleId, m_wNbCycles, m_wInterCycle);

		if (sRet == F28_OK)
		{
			m_ucPhase = EASY_AUTO_WAIT_EOC;
		}
		else
		{
			m_wError = m_ucPhase;
			m_ucPhase = EASY_AUTO_END;
		}
		break;
	case EASY_AUTO_WAIT_EOC:		// Wait EOC
		if (F28_GetEOCRatio(m_sModuleId) > 0)
		{
			m_wError = 0;			// Pas d'erreur
			m_ucPhase = EASY_AUTO_END;
		}
		break;

	case EASY_AUTO_END:					// End of easy auto learning
		m_wError = m_ucPhase;
		m_ucPhase = EASY_AUTO_IDDLE;
		m_bRunning = FALSE;
		bReturn = TRUE;
		break;

	case EASY_AUTO_IDDLE:				// Ready do nothing

		// Do nothing
		break;
	}

	return bReturn;
}
